plugins/ssr/ssrfwd.ml

```diff
@@ -318,3 +318,130 @@ let sufftac ist ((((clr, pats),binders),simpl), ((_, c), hint)) =
     let _,ty,_,uc = pf_interp_ty ist gl c in let gl = pf_merge_uc uc gl in
     basecuttac "ssr_suff" ty gl in
   Tacticals.tclTHENS ctac [htac; Tacticals.tclTHEN (old_cleartac clr) (introstac (binders@simpl))]
+
+open Proofview.Notations
+
+let is_app_evar sigma t =
+  match EConstr.kind sigma t with
+  | Constr.Evar _ -> true
+  | Constr.App(t,_) ->
+      begin match EConstr.kind sigma t with
+      | Constr.Evar _ -> true
+      | _ -> false end
+  | _ -> false
+
+let tmp = ref 0
+
+let rec intro_lock names = Proofview.Goal.enter begin fun gl ->
+ let c = Proofview.Goal.concl gl in
+ let sigma = Proofview.Goal.sigma gl in
+ let env = Proofview.Goal.env gl in
+ let rec aux c =
+  match EConstr.kind_of_type sigma c with
+  | Term.SortType _ -> Proofview.tclUNIT ()
+  | (Term.ProdType (name,_,t) | Term.LetInType (name,_,_,t)) ->
+      begin match names with
+      | n :: names -> Ssripats.(tclIPAT @@ [Ssripats.IOpId n]) <*> intro_lock names
+      | [] ->
+          (* after quell intro patterns, use IpatTemp *)
+          let id =
+            match name with
+            | Name.Anonymous -> Id.of_string (Printf.sprintf "_under_%d_" !tmp)
+            | Name.Name n -> n in
+          incr tmp;
+          Ssripats.tclIPAT [
+            Ssripats.IOpId id;
+            Ssripats.IOpEqGen (intro_lock names);
+            Ssripats.IOpEqGen (Tactics.revert [id])]
+      end
+  | Term.CastType (t,_) -> aux t
+  | Term.AtomicType _ ->
+      let t = Reductionops.whd_all env sigma c in
+      match EConstr.kind_of_type sigma t with
+      | Term.ProdType _ -> aux t
+      | Term.AtomicType(hd, args) when
+          Ssrcommon.is_ind_ref sigma hd (Coqlib.lib_ref "core.eq.type") &&
+          Array.length args = 3 && is_app_evar sigma args.(2) ->
+            Tactics.New.refine ~typecheck:true (fun sigma ->
+              let sigma, under =
+                Ssrcommon.mkSsrConst "Under" env sigma in
+              let sigma, under_from_eq =
+                Ssrcommon.mkSsrConst "Under_from_eq" env sigma in
+              let ty = EConstr.mkApp (under,args) in
+              let sigma, t = Evarutil.new_evar env sigma ty in
+              sigma, EConstr.mkApp(under_from_eq,Array.append args [|t|]))
+      | _ -> Proofview.tclUNIT ()
+  in
+    aux c
+
+end
+
+let rec pretty_rename evar_map term = function
+  | [] -> term
+  | varname :: varnames ->
+     try begin match EConstr.destLambda evar_map term with (_, types, body) ->
+           let res = pretty_rename evar_map body varnames in
+           let name = Names.Name.mk_name varname in
+           (* Note: we don't explicitly check name \notin free_vars(res) here *)
+           EConstr.mkLambda (name, types, res)
+         end
+     with
+     | DestKO ->
+        ppdebug(lazy Pp.(str"under: cannot pretty-rename all bound variables with destLambda"));
+        term
+
+let overtac gl = ssr_n_tac "over" ~-1 gl
```

Comment:
Move here the cast from the old type of tactics to the new one, so that the .mli we export the new type.

Reply:
Done.